doc: makeStyles documentation (#17726)
microsoft/fluentui: packages/react-examples/src/react-components/FluentUIComponents/StylingComponents.stories.mdx

import { Meta, Source } from '@storybook/addon-docs/blocks';

<Meta title="Concepts/Styling Components" />

## Styling components

To style Fluent UI components `makeStyles` is used.

> `makeStyles` is a proprietary CSS-in-JS implementation which generates atomic CSS classes.

Reviewer: ```suggestion
`makeStyles` is a proprietary CSS-in-JS implementation which generates atomic CSS classes and can be imported from "@fluentui/react-components".
```

?

Reviewer 2: ```suggestion
`makeStyles` is a proprietary CSS-in-JS implementation which generates atomic CSS classes.

> import { makeStyles } from '@fluentui/react-components'
```

> To style a component, you need to call `makeStyles` in a module scope to create a hook to be used inside a component.

Reviewer: ```suggestion
To style a component, you need to call `makeStyles` in a module scope to create a React hook to be used inside a component.
```

_nit_

The `makeStyles` call accepts an object of items where each key is a unique identifier and each value is an object with styles.

<!-- TODO: format as <Code /> -->

```jsx

> const useStyles = makeStyles({
>   root: { color: 'red' },

Reviewer: ```suggestion
import { makeStyles } from '@fluentui/react-components';

const useStyles = makeStyles({
  root: { color: 'red' },
```

Let's clarify in an example from where it should be imported.

The same approach is used to apply classes passed from parent component. Again, you need to use `ax()` to properly merge and deduplicate the classes.

```jsx
function Component(props) {
  const classes = useStyles();

  return <div className={ax(classes.root, props.className /* these definitions have higher precedence */)} />;
}
```

### Applying theme to styles

No matter what theme is used, the component styles are always the same. The only way how to change the component appearance is through tokens which can be used in style values.

```js
const useStyles = makeStyles({
  root: { display: 'flex' },
  rootPrimary: (theme) => ({ color: theme.alias.color.neutral.neutralForeground3 }),
});
```

Those tokens are resolved to CSS variable usages. `ThemeProvider` component is responsible for setting the CSS variables in DOM and changing them when theme changes. When theme is switched, only the variables are changed, all styles remain the same.

## Overriding FUI component styles

To override an appearance of a FUI component, you use the exactly same approach - You call makeStyles/useStyles in your code and pass the resulting classes through `props`.

> ```
> const useButtonOverrides = makeStyles({
>   rootPrimary: theme => ({ color: theme.alias.color.neutral.neutralForeground3 }),
> })
> 
> function MyComponent() {
>   const buttonOverrides = useButtonOverrides();
> 
>   return <Button className={buttonOverrides.root} />;
> }
> 

Reviewer: ```suggestion
```jsx
const useOverrides = makeStyles({
  button: theme => ({ color: theme.alias.color.neutral.neutralForeground3 }),
  link: { fontWeight: 'bold },
})

function MyComponent() {
  const overrides = useOverrides();

  return (
    <>
      <Button className={overrides.button} />
      <Link className={overrides.link} />
    </>
  );
}
```

nit, I think that in this case we will better show how to use overrides for multiple components

import { Meta, Source } from '@storybook/addon-docs/blocks';

<Meta title="Concepts/Styling Components" />

## Styling components

To style Fluent UI components `makeStyles` is used.

> `makeStyles` is a proprietary CSS-in-JS implementation which generates atomic CSS classes and can be imported from "@fluentui/react-components".

Reviewer 2: nit: allow easy ctrl+c ctrl+v

```suggestion
`makeStyles` is a proprietary CSS-in-JS implementation which generates atomic CSS classes. Get started by simply importing

> import { makeStyles } from '@fluentui/react-components'
```

> `makeStyles` is a proprietary CSS-in-JS implementation which generates atomic CSS classes and can be imported from "@fluentui/react-components".

Reviewer 2: how is `makeStyles` proprietary ?, do you mean 'our custom implementation' ?

Author: I mean it's not 3rd party OSS library

Reviewer 2: Let's settle with smth this, since proprietary seems to me that we'll get the lawyers involved if you fork it :)

> homegrown CSS-in-JS implementation (inspired by stylex?)

To style a component, you need to call `makeStyles` in a module scope to create a React hook to be used inside a component.

> The `makeStyles` call accepts an object of items where each key is a unique identifier and each value is an object with styles.

Reviewer 2: ```suggestion
The `makeStyles` call accepts an object of items where each key is a unique identifier and each value is an object with styles. The call returns an object with classes mapped to these unique identifiers.
```

<!-- TODO: format as <Code /> -->

```jsx
import { makeStyles } from '@fluentui/react-components';

const useStyles = makeStyles({
  root: { color: 'red' },
});

function Component() {
  const classes = useStyles();

  return <div className={classes.root} />;
}
```

### Merging component styles

> There are cases where you need to merge classes from multiple `useStyles` calls. **It is not possible to just concatenate the classes!**

Reviewer 2: Let's make this more obvious ? 

```suggestion

There are cases where you need to merge classes from multiple `useStyles` calls. 

#### ⚠It is not possible to simply concatenate `useStyles` classnames
```

To properly merge the classes, you need to use `ax()` function, which performs merge and deduplication of atomic classes generated by `makeStyles()`.

> When `ax()` is called, it merges all classes from top to bottom - the latter argument overwrites the previous ones (similar to `Object.assign()`).

Reviewer 2: ```suggestion
When `ax()` is called, it merges all classes from first to last - the latter argument overwrites the previous ones (similar to `Object.assign()`).
```

```jsx
const useStyles = makeStyles({
  blueBold: {
    color: 'blue',
    fontWeight: 'bold',
  },
  red: {
    color: 'red',
  },
});

function Component() {
  const classes = useStyles();

>   const first = ax(classes.blueBold, classes.red); // { color: 'red', fontWeight: 'bold' }

Reviewer 2: I thought we were supposed to use `useAx` now ? 
```suggestion
  const first = useAx(classes.blueBold, classes.red); // { color: 'red', fontWeight: 'bold' }
```

Author: Not anymore - #17713.

  const second = ax(classes.red, classes.blueBold); // { color: 'blue', fontWeight: 'bold' }
}
```

### Mapping props and state to styles

> Both `makeStyles` and `ax` are props and state agnostic. It is developers responsibility to, based on props and state, merge and apply appropriate parts of the `useStyles` call:

Reviewer 2: ```suggestion
Both `makeStyles` and `ax` are props and stateless. It is developers responsibility to, based on props and state, merge and apply appropriate parts of the `useStyles` call:
```

> Both `makeStyles` and `ax` are props and state agnostic. It is developers responsibility to, based on props and state, merge and apply appropriate parts of the `useStyles` call:

Reviewer 2: nit: simpler verbiage ? 

```suggestion
Both `makeStyles` and `ax` stateless and unaware of props. It is developers responsibility to, based on props and state, merge and apply appropriate parts of the `useStyles` call:
```

Reviewer: I suggest to avoid "stupid" and find another word as it's offensive...

Author: Can we go with "simple"?

Reviewer 2: if you want to be boring ? 😛 

Sounds good

```jsx
const useStyles = makeStyles({
  root: { color: 'red' },
  rootPrimary: { color: 'blue' },
});

function Component(props) {
  const classes = useStyles();

  return <div className={ax('ui-component', classes.root, props.primary && classes.rootPrimary)} />;
}
```

### Applying classes passed from parent

> The same approach is used to apply classes passed from parent component. Again, you need to use `ax()` to properly merge and deduplicate the classes.

Reviewer 2: ```suggestion
The same approach is used to apply classes passed from parent component. Again, you need to use `ax()` to properly merge and deduplicate the classes. Note again, that the order of application is important in `ax()` so that parent classes overwrite component classes.
```

```jsx
function Component(props) {
  const classes = useStyles();

>   return <div className={ax(classes.root, props.className /* these definitions have higher precedence */)} />;

Reviewer 2: Instead of the comment, I think it would be clearer to have a bad example of not respecting order

Author: I intentionally did not add the bad example to avoid copy&pasting it :-)

Reviewer 2: Put it in some comments ? The results are interesting to be aware of IMO

Author: Added Incorrect usages section

}
```

### Applying theme to styles

> No matter what theme is used, the component styles are always the same. The only way how to change the component appearance is through tokens which can be used in style values.

Reviewer 2: ```suggestion
No matter what theme is used, the component styles are always the same. The only way how override component styling is through theme tokens which can be used in style values.
```

```js
const useStyles = makeStyles({
  root: { display: 'flex' },
  rootPrimary: (theme) => ({ color: theme.alias.color.neutral.neutralForeground3 }),
});
```

Those tokens are resolved to CSS variable usages. `ThemeProvider` component is responsible for setting the CSS variables in DOM and changing them when theme changes. When theme is switched, only the variables are changed, all styles remain the same.

## Overriding FUI component styles

> To override an appearance of a FUI component, you use the exactly same approach - You call makeStyles/useStyles in your code and pass the resulting classes through `props`.

Reviewer 2: should we add an example for shorthand slots too ? `icon` slot on `Button` should serve this purpose nicely

Author: Icon overrides added

import { Meta, Source } from '@storybook/addon-docs/blocks';

<Meta title="Concepts/Styling Components" />

## Styling components

To style Fluent UI components `makeStyles` is used.

`makeStyles` is a proprietary CSS-in-JS implementation which generates atomic CSS classes. Get started by simply importing

> > import { makeStyles } from '@fluentui/react-components
> > '

Reviewer: ```suggestion
> import { makeStyles } from '@fluentui/react-components'
```

Typo?
